File: web/pgadmin/utils/dynamic_registry/tests/registry/registry.py
#######################################################################
#
# pgAdmin 4 - PostgreSQL Tools
#
# Copyright (C) 2013 - 2024, The pgAdmin Development Team
# This software is released under the PostgreSQL Licence
#
##########################################################################
from pgadmin.utils.dynamic_registry import create_registry_metaclass


TestModuleRegistry = create_registry_metaclass(
    'TestModuleRegistry', __package__, decorate_as_module=True
)


TestNamedRegistry = create_registry_metaclass(
    'TestRegistry', __package__, decorate_as_module=False
)


class TestModuleBase(metaclass=TestModuleRegistry):
    pass


class TestNameBase(metaclass=TestNamedRegistry):
    pass


class ClassTestName1(TestNameBase):
    pass


class ClassTestName2(TestNameBase):
    pass

File: web/pgadmin/static/js/SchemaView/registry.js
/////////////////////////////////////////////////////////////
//
// pgAdmin 4 - PostgreSQL Tools
//
// Copyright (C) 2013 - 2024, The pgAdmin Development Team
// This software is released under the PostgreSQL Licence
//
//////////////////////////////////////////////////////////////

/*
 * Using the factory pattern (registry) to avoid circular imports of the views.
 */
const _views = {};

export function registerView(viewFunc, name) {
  name = name || viewFunc.name;

  if (name in _views) {
    throw new Error(
      `View type '${name}' is alredy registered.`
    );
  }

  if (typeof viewFunc !== 'function') {
    throw new Error(
      `View '${name}' must be a function.`
    );
  }

  _views[name] = viewFunc;
}

export function View(name) {
  const view = _views[name];

  if (view) return view;
  throw new Error(`View ${name} is not found in the registry.`);
}

export function hasView(name) {
  return (name in _views);
}
